### PR TITLE
A11y metatip color swatches (+ click to copy to clipboard)

### PR DESCRIPTION
**Pull request by hchiam**
For #496 
demo-1 
As a simple fix to side-step contrast issues, I went with a text outline, instead of using `mix-blend-mode: difference` (which i think could end up with gray on gray). if you'd like, i can look further into how pika dynamically shows just black xor white text depending on the swatch color, which would require fancier js logic. Maybe i could also add some js logic to make the "copied!" message stay a bit longer. lemme know your thoughts! First time contributing to VisBug :)

### CODE REVIEW

File: app/features/accessibility.js
Comment (argyleink) [inline app/features/accessibility.js:181]:
```suggestion
          <span>${new TinyColor(background)[colormode]()}</span>
```

**Review by argyleink — CHANGES_REQUESTED** [review]
this is rad work! i def think we can get this merged without too much more effort. i really appreciate you hacking on this, it looks so much better already.

a minor thing is presenting the color based on the user preference. i've added some inline comments to help out.

the biggest thing to fix / think about / design (🙂 ), is when inspecting elements with aria roles and other a11y meta, the new color boxes shows their fixed size (see screenshot 1). to solve this and match pika, i think separating the score from this new visual would make the layout easier to handle being responsive. i think having your new color block side by side is best above the `<code accessibility>` element in `ally.element.js`.

image

Would be nice to have it be responsive too, so on mobile or testing mobile, the layout fits into a portrait layout 

image

Is gitter or slack a better way to have some direct conversation? I want to help you and save you time by assisting you quickly 🙂 I wasnt able to quickly have a design solution locally, i think the templates need some attention. we're deviating from the metatip formula, so we'll need to teach it how to present these new elements! no prob, we can totally get there.

besides those 2 things, are you planning on matching Pika further by changing the way `AA, AA+, AAA, AAA+` look? 😁

**Comment by argyleink** [comment]
oh yeah! and tinycolor is available. you could use it's api to find out if white/black are better (so you dont need all the text shadows), here's the api https://www.npmjs.com/package/@ctrl/tinycolor. you might even be able to get more creative than than black/white, it's a powerful color tool 👍🏻

File: app/components/metatip/metatip.element.css
Comment (hchiam) [reply app/components/metatip/metatip.element.css:258]: unfortunately it didn't work on my screen :(

File: app/features/accessibility.js
Comment (hchiam) [reply app/features/accessibility.js]: nice! 
image

**Comment by hchiam** [comment]
demo-2

the other 2 things i'll need to look into are simply rearranging the AA, AA+, AAA, AAA+ section and hopefully fix the one failing test

File: app/components/metatip/metatip.element.css
Comment (argyleink) [reply app/components/metatip/metatip.element.css:279]: i see now too, it's swapping content and there's no way to delay that. might need 2 separate elements! cross fade between the 2?

File: app/utilities/accessibility.js
Comment (argyleink) [inline app/utilities/accessibility.js:58]: I forgot to mention [I wrote one here](https://github.com/GoogleChromeLabs/ProjectVisBug/blob/main/app/features/color.js#L176), for the icons overlaying the colors in the sidebar, but I like the use of `mostReadable()` here. When i tried this out locally, sometimes pretty odd colors would be chosen, like the random or the inversion/180 rotate, which produced some pretty bad combos. I updated it to this, which felt more stable:

```suggestion
export const getContrastingColor = color => {
  return mostReadable(
    color,
    ['#000', '#fff'],
    {
      includeFallbackColors: true,
      level: 'AAA',
      size: 'small'
    }
  )
```

but i also didnt hate using the existing one with a tweak for more contrast. Sometimes it picks a color in the same family as what it's displaying, which looks really nice. and it clips to black/white:

```js
  let contrast = new TinyColor(color)

  contrast = contrast.isDark()
    ? contrast.lighten(75)
    : contrast.darken(75)

  return contrast
```

fun! i don't know what's best yet 🙂

File: app/components/metatip/ally.element.js
Comment (argyleink) [inline app/components/metatip/ally.element.js:72]: there's an empty div getting made here if not hovering something with aria roles or other screen reader information. try hovering on some of the form inputs on the demo page.

File: app/components/metatip/ally.element.js
Comment (hchiam) [reply app/components/metatip/ally.element.js:20]: sounds good, i like the idea of a util function with fallback. i'll use your suggestion for now for consistency/easier-searchability

File: extension/manifest.json
Comment (hchiam) [reply extension/manifest.json:11]: good news: i tested `clipboard.writeText` without `clipboard-write` in firefox and it worked for me, so we don't need this permission anymore

File: app/utilities/accessibility.js
Comment (hchiam) [reply app/utilities/accessibility.js:58]: btw i read that when you have `includeFallbackColors: true` set, then `mostReadable` will automatically fall back to black or white if none of the items in the array param have good enough contrast, so i think i actually prefer your existing function

Comment (hchiam) [reply app/utilities/accessibility.js:58]: actually, I think I'll use your suggestion above. i tried the tweak on the existing one and sometimes got low contrast (i'm in dark mode). 
Screen Shot 2021-05-12 at 20 10 58

File: app/components/metatip/ally.element.js
Comment (hchiam) [reply app/components/metatip/ally.element.js]: i made the empty div disappear now - and the 10px gap in the actual presentation

File: app/components/metatip/metatip.element.css
Comment (hchiam) [reply app/components/metatip/metatip.element.css:278]: copied-message-persists

powered by `tabindex="0"` (makes any element focusable and hence can "hold" onto focus after you click) 

If you'd like the "copied!" to stay even after you hover away, i can add this one line:
```suggestion
  opacity: 1;
  content: "📋 copied!";
```

**Review by hchiam — COMMENTED** [review]
ready for one more check 😃 @argyleink https://github.com/GoogleChromeLabs/ProjectVisBug/pull/505#discussion_r631496577

File: app/components/metatip/metatip.element.css
Comment (argyleink) [reply app/components/metatip/metatip.element.css:278]: works great as is! 👍🏻

File: app/components/metatip/metatip.element.css
Comment (argyleink) [inline app/components/metatip/metatip.element.css:241]:
```suggestion
  padding: 10px;
```

this aligns the color-swatch content with the accessibility content, was just a hair off when using ems in this spot. screenshot shows the result of switching to pixels 👍🏻 

Screen Shot 2021-05-13 at 11 45 30 AM

**Comment by argyleink** [comment]
It's lookin so good! There's 2 remaining questions/tasks before this merges into main (which will cause a build and deploy to the app store!)
1. ~match the presentation that pika has for the ratio scores. I put ~ in there because I don't think we need to 100% rip the design, like right now visbug is smart enough to choose between AA or AAA. We should keep that, but adopt nice parts of Pika's. thoughts? 
2. figure out how to show aria information and other accessibility meta data that visbug scrapes from elements. it should probably go under the contrast scores? it's an open design task.

you feeling up to these last items? you've done a great set of work here and I don't want to exhaust you! 🙂  how you feelin? 

if it's helpful, i could create a design or 2 to show the direction i'm thinking (solve the design task), creating a focused engineering task?

**Comment by hchiam** [comment]
no worries, it's been fun! a design/mock would be helpful to make sure I'm on track. but yeah, feel free to make a separate branch that i can merge this PR into instead of main, for you to work on.  meanwhile i'll think about the design and see how far i get

**Comment by hchiam** [comment]
sizing+order+alignment

**Comment by argyleink** [comment]
you've done so much, it looks sooo good 🙂  
when you're done, let me know and I'll merge this! I'll create a new PR once I've rounded out some of those remaining design items 👍🏻 

I've updated this PRs base to merge into an "ally metatip v2" branch, so i/we can collaborate a bit easier and get it perfect before merging into `main` 💯

**Comment by hchiam** [comment]
thanks! Sounds good, go ahead

**Pull request merged**
